greeclimate/network.py

```diff
@@ -95,16 +117,39 @@ def resume_writing(self) -> None:
         self._drained.set()
         super().resume_writing()
 
+    @staticmethod
+    def decrypt_payload(payload, key=GENERIC_KEY[0]):
+        cipher = AES.new(key.encode(), AES.MODE_ECB)
+        decoded = base64.b64decode(payload)
+        decrypted = cipher.decrypt(decoded).decode()
+        t = decrypted.replace(decrypted[decrypted.rindex("}") + 1 :], "")
+        return json.loads(t)
+
+    @staticmethod
+    def encrypt_payload(payload, key=GENERIC_KEY[0]):
+        def pad(s):
+            bs = 16
+            return s + (bs - len(s) % bs) * chr(bs - len(s) % bs)
+
+        cipher = AES.new(key.encode(), AES.MODE_ECB)
+        encrypted = cipher.encrypt(pad(json.dumps(payload)).encode())
+        encoded = base64.b64encode(encrypted).decode()
+
+        _LOGGER.debug(f"Encrypted payload with key [{key}]: {encoded}")
+        return encoded
+
     def datagram_received(self, data: bytes, addr: IPAddr) -> None:
         """Handle an incoming datagram."""
         if len(data) == 0:
             return
 
         obj = json.loads(data)
-        key = GENERIC_KEY if obj.get("i") == 1 else self._key
+
+        # It could be either a v1 or v2 key
+        key = GENERIC_KEY[0] if obj.get("i") == 1 else self._key
```

Comment:
Sorry to interrupt your coding flow, but I can see that you incorporating the new KEY and encoding for new models with firmware 1.2.1. Would you mind if I through in a few suggestions. I have a new Gree AC and I would like to be able to control it using HA.

Reply:
The plan is to add the GCM encryption support following this change, it's not actually implemented yet though.

To be pedantic. V1.2.1 sounds like you're referring to the hardware revision, my device reports 1.2.1 and uses the older encryption technique. We won't be able to select based on that version reported in the scan result.

Reply:
V1.2.1 is the firmware version showing on Gree app. 
The best approach is to try the GCM if the current encryption causes a timeout. 
If both make a timeout then we raise a timeout.

Reply:
Interesting... they have a weird concept of versions, but yes. Timeout approach is what I was thinking too.

That said, could you share responses from the scan and bind commands from your 1.2.1 devices please? Maybe there is something in there, I can compare with examples I have on hand.

Reply:
Here is the response from the scan 
```json
{"t": "pack", "i": 1, "uid": 0, "cid": "", "tcid": "", "pack": {"t": "dev", "cid": "9424b8bb14bc", "bc": "00000000000000000000000000000000", "brand": "gree", "catalog": "gree", "mac": "9424b8bb14bc", "mid": "10001", "model": "gree", "name": "", "lock": 0, "series": "gree", "vender": "1", "ver": "V2.0.0", "ModelType": "32772", "hid": "362001065279+U-WB05RT13V1.21.bin"}}
```
hid is showing `V1.21`